src/verify.rs

```diff
@@ -1,43 +1,112 @@
+        "id-sha512" => Some(Box::new(sha2::Sha512::new())),
+        "id-sha224" => Some(Box::new(sha2::Sha224::new())),
+        _ => None,
     }
 }
 
-pub fn check_against_attribute_certificate(pe: &PE, certificate: &AttributeCertificate) -> bool {
-    match certificate.as_spc_indirect_data_content() {
-        Some(Ok(spc_data)) => {
-            // derive dyndigest from digest_algorithm…
-            if let Some(hasher) = dyndigest_from_algo_identifier(spc_data.message_digest.digest_algorithm) {
-                let output = pe.authenticode_dyndigest(hasher);
+pub fn verify_pe_signatures(pe: &PE, options: VerificationOptions) -> Result<(bool, Vec<VerificationError>), VerificationError> {
+    let certificates = certificates_from_pe(pe);
+    let mut hashes = HashMap::new();
+    let mut verified = 0;
+    let mut errors = Vec::new();
+
+    if certificates.is_empty() {
+        return Err(VerificationError::NoCertificate);
+    }
+
+    for (sdata, message_digest) in certificates {
+        if let Some(hasher) = get_hasher(&message_digest.digest_algorithm) {
+            let authenticode = hashes
+                .entry(message_digest.digest_algorithm.oid)
+                .or_insert(pe.authenticode_dyndigest(hasher));
+
+            if sdata.certificates.is_none() {
+                errors.push(VerificationError::MissingCertificate(sdata.signer_infos));
+            }
+
+            let valid_authenticode = message_digest.digest.as_bytes() == &authenticode[..];
+
+            if !valid_authenticode {
+                errors.push(VerificationError::InvalidAuthenticode(sdata.certificates.clone(), format!("{:x?}", authenticode),
+                    format!("{:x?}", message_digest.digest.as_bytes())));
+            }
 
-                &output[..] == spc_data.message_digest.digest.as_bytes()
-            } else {
-                false
+            // Verify if any of the certificate is valid among a set of trusted certificates.
+            // If there is no trusted certificates in the trust store, we skip this test.
+            let trusted_certificates = match options.trust_store {
+                Some(ref trusted_certs) => trusted_certs.iter().map(|trusted_cert|
+                    VerifyingKey::try_from(trusted_cert)
+                ).collect::<Result<Vec<_>, _>>().expect("Failed to transform a trusted certificate into a verifying key; unsupported certificate?"),
+                None => Vec::new()
+            };
+            let mut any_valid_certificate = trusted_certificates.is_empty();
+            for certificate in sdata.certificates.map(|certs| certs.0.into_vec()).unwrap() {
+                match certificate {
+                    cms::cert::CertificateChoices::Certificate(certificate) => {
+                        any_valid_certificate = any_valid_certificate ||
+                            trusted_certificates.iter().any(|trusted_cert| trusted_cert.verify(&certificate).is_ok());
```

Comment:
I should enqueue verification failures here and dispatch any unexpected error too.